# Image preprocessing

----

This notebook can be used to preprocess the illumination corrected raw images obtained from [IDR0033](https://idr.openmicroscopy.org/webclient/?show=screen-1751) for the further analysis.

We will use available metadata to filter images that were manually identified by the authors of the [publication](https://elifesciences.org/articles/24060#s4) corresponding to the published imaging data as being outliers or that did not pass their quality control steps. For more information concerning the applied workflows to identify those, please refer to the publication.

---

## 0. Environmental setup

In [1]:
import os
import sys
import pandas as pd
import numpy as np
import copy
import matplotlib.pyplot as plt
import tifffile

from collections import Counter
from tqdm.notebook import tqdm
from shutil import copyfile
from skimage import (
    filters,
    morphology,
    measure,
    segmentation,
    exposure,
    feature,
    io,
    color,
)
import scipy.ndimage as ndi


sys.path.append("..")

from src.utils.basic.io import get_file_list
from src.utils.basic.visualization import (
    explore_samples_of_dataset,
    explore_segmentation_of_dataset,
)

# Load automatic code formatter (automatically executed when running a chunk)
%load_ext nb_black

<IPython.core.display.Javascript object>

In [2]:
def rename_image_filenames(
    metadata,
    orig_col="Image_FileName_OrigHoechst",
    illum_col="Image_FileName_IllumHoechst",
    posfix="illum_corrected",
):
    orig_image_file_names = list(metadata[orig_col])
    illum_corrected_image_file_names = []
    for orig_image_file_name in orig_image_file_names:
        idx = orig_image_file_name.index(".")
        illum_corrected_image_file_name = (
            orig_image_file_name[:idx] + posfix + orig_image_file_name[idx:]
        )
        illum_corrected_image_file_names.append(illum_corrected_image_file_name)
    metadata[illum_col] = illum_corrected_image_file_names
    return metadata

<IPython.core.display.Javascript object>

In [3]:
def filter_out_qc_flagged_items(
    metadata,
    blurry_col="Image_Metadata_QCFlag_isBlurry",
    saturated_col="Image_Metadata_QCFlag_isSaturated",
):
    filtered_metadata = metadata.loc[
        (metadata[blurry_col] == 0) & (metadata[saturated_col] == 0)
    ]
    return filtered_metadata

<IPython.core.display.Javascript object>

In [4]:
def remove_outlier_items(
    metadata,
    outlier_plates=None,
    outlier_plate_wells=None,
    outlier_wells=None,
    plate_col="Image_Metadata_Plate",
    well_col="Image_Metadata_Well",
):
    metadata_orm = metadata.copy()
    if outlier_plates is not None:
        for outlier_plate in outlier_plates:
            metadata_orm = metadata_orm.loc[metadata_orm[plate_col] != outlier_plate]
    if outlier_plate_wells is not None:
        for outlier_plate_well in outlier_plate_wells:
            metadata_orm = metadata_orm.loc[
                (metadata_orm[plate_col] != outlier_plate_well[0])
                | (metadata_orm[well_col] != outlier_plate_well[1])
            ]
    if outlier_wells is not None:
        for outlier_well in outlier_wells:
            metadata_orm = metadata_orm.loc[metadata_orm[well_col] != outlier_well]
    return metadata_orm

<IPython.core.display.Javascript object>

In [5]:
def save_filtered_images(
    metadata,
    input_dir,
    output_dir,
    plate_col="Image_Metadata_Plate",
    illum_file_col="Image_FileName_IllumHoechst",
):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    n = len(metadata)
    for i in tqdm(range(n), desc := "Copying images"):
        plate = metadata.iloc[i, list(metadata.columns).index(plate_col)]
        filename = metadata.iloc[i, list(metadata.columns).index(illum_file_col)]

        plate_output_dir = os.path.join(output_dir, str(plate))
        plate_input_dir = os.path.join(input_dir, str(plate))
        if not os.path.exists(plate_output_dir):
            os.makedirs(plate_output_dir)

        plate_output_file = os.path.join(plate_output_dir, filename)
        plate_input_file = os.path.join(plate_input_dir, filename)

        copyfile(plate_input_file, plate_output_file)
    print("Images (n={}) copied to {}.".format(len(metadata), output_dir))

<IPython.core.display.Javascript object>

In [6]:
def create_sample_segmentation_datasets(
    metadata,
    image_input_dir,
    outline_input_dir,
    n=10,
    plate_col="Image_Metadata_Plate",
    illum_file_col="Image_FileName_IllumHoechst",
    nuclei_outline_col="Image_FileName_NucleiOutlines",
    random_state=1234,
):
    sample_segmentation_datasets = {}
    sample_image_datasets = {}
    plates = np.unique(metadata.loc[:, plate_col])
    np.random.seed(random_state)
    for plate in plates:
        plate_segmentation_dataset = []
        plate_image_dataset = []
        plate_metadata = metadata.loc[metadata[plate_col] == plate]
        sample_idc = np.random.choice(len(plate_metadata), n)
        plate_image_dir = os.path.join(image_input_dir, str(plate))
        plate_outline_dir = os.path.join(outline_input_dir, str(plate))
        for idx in sample_idc:
            image_file_name = plate_metadata.iloc[idx, :][illum_file_col]
            image_file_path = os.path.join(plate_image_dir, image_file_name)
            outline_file_name = plate_metadata.iloc[idx, :][nuclei_outline_col]
            outline_file_path = os.path.join(plate_outline_dir, outline_file_name)

            image = io.imread(image_file_path)
            outlines = io.imread(outline_file_path)

            segmented_image = segment_image(outlines)
            plate_segmentation_dataset.append(segmented_image)
            plate_image_dataset.append(image)
        sample_image_datasets[str(plate)] = plate_image_dataset
        sample_segmentation_datasets[str(plate)] = plate_segmentation_dataset
    return sample_image_datasets, sample_segmentation_datasets

<IPython.core.display.Javascript object>

In [7]:
def segment_image(contour_image):
    binary = ndi.binary_fill_holes(contour_image)
    binary[contour_image != 0] = 0
    labeled = measure.label(binary)
    return labeled

<IPython.core.display.Javascript object>

In [92]:
def segment_images_and_get_nuclei_metadata(
    metadata,
    image_input_dir,
    outline_input_dir,
    output_dir,
    plate_col="Image_Metadata_Plate",
    illum_file_col="Image_FileName_IllumHoechst",
    nuclei_outline_col="Image_FileName_NucleiOutlines",
    nuclei_count_col="Image_Count_Nuclei",
    min_area=1,
):
    nuclei_metadata = []
    image_metadata = metadata.copy()

    metadata_cols = list(metadata.columns)
    max_width = 0
    max_length = 0

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    for i in tqdm(range(len(metadata))):
        plate = str(metadata.iloc[i, :][plate_col])

        image_file_name = metadata.iloc[i, :][illum_file_col]
        image_file_path = os.path.join(input_dir, plate, image_file_name)
        image = tifffile.imread(image_file_path)

        outline_file_name = metadata.iloc[i, :][nuclei_outline_col]
        outline_file_path = os.path.join(outline_input_dir, plate, outline_file_name)
        outline_image = io.imread(outline_file_path)

        label_image = segment_image(outline_image)

        label_image = morphology.remove_small_objects(label_image, min_size=min_area)

        image_metadata.iloc[i, metadata_cols.index(nuclei_count_col)] = len(
            np.unique(label_image)
        )

        regions = measure.regionprops(label_image=label_image, intensity_image=image)
        for region in regions:
            width, length = region.image.shape
            max_width = max(max_width, width)
            max_length = max(max_length, length)

            fname_start = image_file_name[: image_file_name.index(".")]
            fname_ending = image_file_name[image_file_name.index(".") :]
            plate_output_dir = os.path.join(output_dir, plate)

            if not os.path.exists(plate_output_dir):
                os.makedirs(plate_output_dir)

            output_file_name = os.path.join(
                plate_output_dir,
                fname_start + "_{}".format(region.label) + fname_ending,
            )

            cropped = region.intensity_image

            tifffile.imsave(output_file_name, cropped)
            nucleus_metadata = list(metadata.iloc[i, :])
            nucleus_metadata[metadata_cols.index(illum_file_col)] = os.path.split(
                output_file_name
            )[1]
            nuclei_metadata.append(nucleus_metadata)
    nuclei_metadata = pd.DataFrame(np.array(nuclei_metadata), columns=metadata_cols)
    nuclei_metadata = nuclei_metadata.loc[
        :,
        [
            "Image_Metadata_Plate",
            "Image_Metadata_Well",
            "Image_FileName_IllumHoechst",
            "Image_Metadata_GeneID",
            "Image_Metadata_GeneSymbol",
            "Image_Metadata_IsLandmark",
            "Image_Metadata_AlleleDesc",
            "Image_Metadata_ExpressionVector",
            "Image_Metadata_FlaggedForToxicity",
            "Image_Metadata_IE_Blast_noBlast",
            "Image_Metadata_IntendedOrfMismatch",
            "Image_Metadata_OpenOrClosed",
            "Image_Metadata_RNAiVirusPlateName",
            "Image_Metadata_Site",
            "Image_Metadata_Type",
            "Image_Metadata_Virus_Vol_ul",
            "Image_Metadata_TimePoint_Hours",
            "Image_Metadata_ASSAY_WELL_ROLE",
        ],
    ]
    nuclei_metadata.columns = [
        "plate",
        "well",
        "image_file",
        "gene_id",
        "gene_symbol",
        "is_landmark",
        "allele",
        "expr_vec",
        "toxicity",
        "ie_blast",
        "intended_orf_mismatch",
        "open_closed",
        "rnai_plate",
        "site",
        "type",
        "virus_vol",
        "timepoint",
        "assay_well_role",
    ]
    print(
        "Nuclei segmentation complete: max dimensions ({}, {})".format(
            max_width, max_length
        )
    )
    return nuclei_metadata, image_metadata

<IPython.core.display.Javascript object>

---

## 1. Read in data

First, we will read in the metadata information that provide information about e.g. which images correspond to which plate, well combination and which gene was targeted. The respective file was exported from a database for which the corresponding sql script has been published alongside with the imaging data. Please refer to the github repo of the original publication for more information of how to set up the database. To derive the metadata file load in in the following, we provide a custom sql-file (`extract_metadata.sql`).

In [9]:
metadata = pd.read_csv("../data/images/metadata/metadata_image_data.csv")
metadata.head()

Image_Metadata_Plate Image_Metadata_Well  \
0                 41744                 k21   
1                 41744                 i13   
2                 41744                 j16   
3                 41744                 m07   
4                 41744                 i04   

                          Image_FileName_OrigHoechst  \
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...   
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...   
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...   
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...   
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...   

  Image_FileName_NucleiOutlines  Image_Count_Nuclei  Image_Metadata_GeneID  \
0     k21_s7_NucleiOutlines.png                  60                 1977.0   
1     i13_s4_NucleiOutlines.png                  69                22943.0   
2     j16_s9_NucleiOutlines.png                  48                22926.0   
3     m07_s5_NucleiOutlines.png                  57                 5045.0   
4     i04_s5_NucleiOutlines.png                  51                 5599.0   

  Image_Metadata_GeneSymbol  Image_Metadata_IsLandmark  \
0                     EIF4E                        0.0   
1                      DKK1                        0.0   
2                      ATF6                        1.0   
3                     FURIN                        0.0   
4                     MAPK8                        0.0   

  Image_Metadata_AlleleDesc Image_Metadata_ExpressionVector  ...  \
0                      WT.2                          pLX304  ...   
1                        WT                          pLX304  ...   
2                      WT.1                          pLX304  ...   
3                      WT.2                          pLX304  ...   
4                      WT.2                          pLX304  ...   

  Image_Metadata_IntendedOrfMismatch  Image_Metadata_OpenOrClosed  \
0                                NaN                         open   
1                                NaN                         open   
2                                NaN                         open   
3                                NaN                         open   
4                                NaN                         open   

  Image_Metadata_RNAiVirusPlateName Image_Metadata_Site  \
0                   ORA11.12.13.18A                   7   
1                   ORA11.12.13.18A                   4   
2                   ORA11.12.13.18A                   9   
3                   ORA11.12.13.18A                   5   
4                   ORA11.12.13.18A                   5   

  Image_Metadata_TimePoint_Hours  Image_Metadata_Type  \
0                            72H               ORF OE   
1                            72H               ORF OE   
2                            72H               ORF OE   
3                            72H               ORF OE   
4                            72H               ORF OE   

  Image_Metadata_Virus_Vol_ul Image_Metadata_ASSAY_WELL_ROLE  \
0                           1                        Treated   
1                           1                        Treated   
2                           1                        Treated   
3                           1                        Treated   
4                           1                        Treated   

   Image_Metadata_QCFlag_isBlurry Image_Metadata_QCFlag_isSaturated  
0                               0                                 0  
1                               0                                 0  
2                               0                                 0  
3                               0                                 0  
4                               0                                 0  

[5 rows x 22 columns]

<IPython.core.display.Javascript object>

As suggested in the original publication we will be working with the image that were corrected for different illumination conditions that is also available.

We will now adapt the `Image_FileName_OrigHoechst` entries in the metadata dataframe by the posfix `_illum_corrected` to ensure that the column represents the actual filenames of `.tif` images that we will be working with.

In [10]:
posfix = "_illum_corrected"
orig_col = "Image_FileName_OrigHoechst"
illum_col = "Image_FileName_IllumHoechst"

metadata = rename_image_filenames(
    metadata, orig_col=orig_col, illum_col=illum_col, posfix=posfix
)
metadata.head()

Image_Metadata_Plate Image_Metadata_Well  \
0                 41744                 k21   
1                 41744                 i13   
2                 41744                 j16   
3                 41744                 m07   
4                 41744                 i04   

                          Image_FileName_OrigHoechst  \
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...   
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...   
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...   
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...   
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...   

  Image_FileName_NucleiOutlines  Image_Count_Nuclei  Image_Metadata_GeneID  \
0     k21_s7_NucleiOutlines.png                  60                 1977.0   
1     i13_s4_NucleiOutlines.png                  69                22943.0   
2     j16_s9_NucleiOutlines.png                  48                22926.0   
3     m07_s5_NucleiOutlines.png                  57                 5045.0   
4     i04_s5_NucleiOutlines.png                  51                 5599.0   

  Image_Metadata_GeneSymbol  Image_Metadata_IsLandmark  \
0                     EIF4E                        0.0   
1                      DKK1                        0.0   
2                      ATF6                        1.0   
3                     FURIN                        0.0   
4                     MAPK8                        0.0   

  Image_Metadata_AlleleDesc Image_Metadata_ExpressionVector  ...  \
0                      WT.2                          pLX304  ...   
1                        WT                          pLX304  ...   
2                      WT.1                          pLX304  ...   
3                      WT.2                          pLX304  ...   
4                      WT.2                          pLX304  ...   

  Image_Metadata_OpenOrClosed  Image_Metadata_RNAiVirusPlateName  \
0                        open                    ORA11.12.13.18A   
1                        open                    ORA11.12.13.18A   
2                        open                    ORA11.12.13.18A   
3                        open                    ORA11.12.13.18A   
4                        open                    ORA11.12.13.18A   

  Image_Metadata_Site Image_Metadata_TimePoint_Hours Image_Metadata_Type  \
0                   7                            72H              ORF OE   
1                   4                            72H              ORF OE   
2                   9                            72H              ORF OE   
3                   5                            72H              ORF OE   
4                   5                            72H              ORF OE   

   Image_Metadata_Virus_Vol_ul Image_Metadata_ASSAY_WELL_ROLE  \
0                            1                        Treated   
1                            1                        Treated   
2                            1                        Treated   
3                            1                        Treated   
4                            1                        Treated   

  Image_Metadata_QCFlag_isBlurry  Image_Metadata_QCFlag_isSaturated  \
0                              0                                  0   
1                              0                                  0   
2                              0                                  0   
3                              0                                  0   
4                              0                                  0   

                         Image_FileName_IllumHoechst  
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...  
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...  
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...  
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...  
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...  

[5 rows x 23 columns]

<IPython.core.display.Javascript object>

---

## 2. Data filtering

### 2a. Filter out blurry or saturated images

Next, we will filter out images that were identified to be blurry or saturated and thus not passing the standards for the image quality that we take over from the authors of the original publication. The respective information are also available in the metadata.

In [11]:
blurry_col = "Image_Metadata_QCFlag_isBlurry"
saturated_col = "Image_Metadata_QCFlag_isSaturated"

filtered_metadata = filter_out_qc_flagged_items(
    metadata, blurry_col=blurry_col, saturated_col=saturated_col
)
filtered_metadata.head()

Image_Metadata_Plate Image_Metadata_Well  \
0                 41744                 k21   
1                 41744                 i13   
2                 41744                 j16   
3                 41744                 m07   
4                 41744                 i04   

                          Image_FileName_OrigHoechst  \
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...   
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...   
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...   
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...   
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...   

  Image_FileName_NucleiOutlines  Image_Count_Nuclei  Image_Metadata_GeneID  \
0     k21_s7_NucleiOutlines.png                  60                 1977.0   
1     i13_s4_NucleiOutlines.png                  69                22943.0   
2     j16_s9_NucleiOutlines.png                  48                22926.0   
3     m07_s5_NucleiOutlines.png                  57                 5045.0   
4     i04_s5_NucleiOutlines.png                  51                 5599.0   

  Image_Metadata_GeneSymbol  Image_Metadata_IsLandmark  \
0                     EIF4E                        0.0   
1                      DKK1                        0.0   
2                      ATF6                        1.0   
3                     FURIN                        0.0   
4                     MAPK8                        0.0   

  Image_Metadata_AlleleDesc Image_Metadata_ExpressionVector  ...  \
0                      WT.2                          pLX304  ...   
1                        WT                          pLX304  ...   
2                      WT.1                          pLX304  ...   
3                      WT.2                          pLX304  ...   
4                      WT.2                          pLX304  ...   

  Image_Metadata_OpenOrClosed  Image_Metadata_RNAiVirusPlateName  \
0                        open                    ORA11.12.13.18A   
1                        open                    ORA11.12.13.18A   
2                        open                    ORA11.12.13.18A   
3                        open                    ORA11.12.13.18A   
4                        open                    ORA11.12.13.18A   

  Image_Metadata_Site Image_Metadata_TimePoint_Hours Image_Metadata_Type  \
0                   7                            72H              ORF OE   
1                   4                            72H              ORF OE   
2                   9                            72H              ORF OE   
3                   5                            72H              ORF OE   
4                   5                            72H              ORF OE   

   Image_Metadata_Virus_Vol_ul Image_Metadata_ASSAY_WELL_ROLE  \
0                            1                        Treated   
1                            1                        Treated   
2                            1                        Treated   
3                            1                        Treated   
4                            1                        Treated   

  Image_Metadata_QCFlag_isBlurry  Image_Metadata_QCFlag_isSaturated  \
0                              0                                  0   
1                              0                                  0   
2                              0                                  0   
3                              0                                  0   
4                              0                                  0   

                         Image_FileName_IllumHoechst  
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...  
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...  
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...  
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...  
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...  

[5 rows x 23 columns]

<IPython.core.display.Javascript object>

In [12]:
print(
    "Images filtered out for not passing the quality standards: {}.".format(
        len(metadata) - len(filtered_metadata)
    )
)

Images filtered out for not passing the quality standards: 251.


<IPython.core.display.Javascript object>

As seen above 251 images were identified to be either blurry or saturated and thus should be excluded for the downstream analysis.

---

### 2b. Filter out outlier images (manual selection)

In addition to the images that were flagged for not passing the quality control steps the authors further excluded 2 additional plate-well combinations and one complete plate during their analyses as they identified those by visual inspection as outliers.

Those are the following:
* Plate 41749 (all wells)
* Plate 41754 (well B01)
* Plate 41757 (well E17)

Unfortunately, no description is given which criteria was used to identify these outliers. When briefly looking at the data in the [IDR webclient](https://idr.openmicroscopy.org/webclient/?show=screen-1751) we do not see any remarkable abnormalities.

Nonetheless, we derive the subset of the dataset where we filter out the corresponding items of these plate-well combinations to follow the preprocessing steps of the original publication. If we will use this subset or the larger set that includes those combinations in our final study is yet to be determined.

In [13]:
outlier_plates = [41749]
outlier_plate_wells = [[41754, "b01"], [41757, "e17"]]
outlier_wells = []
plate_col = "Image_Metadata_Plate"
well_col = "Image_Metadata_Well"


filtered_metadata_orm = remove_outlier_items(
    metadata,
    outlier_plates=outlier_plates,
    outlier_plate_wells=outlier_plate_wells,
    outlier_wells=outlier_wells,
    plate_col=plate_col,
    well_col=well_col,
)
filtered_metadata_orm.head()

Image_Metadata_Plate Image_Metadata_Well  \
0                 41744                 k21   
1                 41744                 i13   
2                 41744                 j16   
3                 41744                 m07   
4                 41744                 i04   

                          Image_FileName_OrigHoechst  \
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...   
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...   
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...   
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...   
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...   

  Image_FileName_NucleiOutlines  Image_Count_Nuclei  Image_Metadata_GeneID  \
0     k21_s7_NucleiOutlines.png                  60                 1977.0   
1     i13_s4_NucleiOutlines.png                  69                22943.0   
2     j16_s9_NucleiOutlines.png                  48                22926.0   
3     m07_s5_NucleiOutlines.png                  57                 5045.0   
4     i04_s5_NucleiOutlines.png                  51                 5599.0   

  Image_Metadata_GeneSymbol  Image_Metadata_IsLandmark  \
0                     EIF4E                        0.0   
1                      DKK1                        0.0   
2                      ATF6                        1.0   
3                     FURIN                        0.0   
4                     MAPK8                        0.0   

  Image_Metadata_AlleleDesc Image_Metadata_ExpressionVector  ...  \
0                      WT.2                          pLX304  ...   
1                        WT                          pLX304  ...   
2                      WT.1                          pLX304  ...   
3                      WT.2                          pLX304  ...   
4                      WT.2                          pLX304  ...   

  Image_Metadata_OpenOrClosed  Image_Metadata_RNAiVirusPlateName  \
0                        open                    ORA11.12.13.18A   
1                        open                    ORA11.12.13.18A   
2                        open                    ORA11.12.13.18A   
3                        open                    ORA11.12.13.18A   
4                        open                    ORA11.12.13.18A   

  Image_Metadata_Site Image_Metadata_TimePoint_Hours Image_Metadata_Type  \
0                   7                            72H              ORF OE   
1                   4                            72H              ORF OE   
2                   9                            72H              ORF OE   
3                   5                            72H              ORF OE   
4                   5                            72H              ORF OE   

   Image_Metadata_Virus_Vol_ul Image_Metadata_ASSAY_WELL_ROLE  \
0                            1                        Treated   
1                            1                        Treated   
2                            1                        Treated   
3                            1                        Treated   
4                            1                        Treated   

  Image_Metadata_QCFlag_isBlurry  Image_Metadata_QCFlag_isSaturated  \
0                              0                                  0   
1                              0                                  0   
2                              0                                  0   
3                              0                                  0   
4                              0                                  0   

                         Image_FileName_IllumHoechst  
0  taoe005-u2os-72h-cp-a-au00044859_k21_s7_w10efe...  
1  taoe005-u2os-72h-cp-a-au00044859_i13_s4_w13be2...  
2  taoe005-u2os-72h-cp-a-au00044859_j16_s9_w1b03e...  
3  taoe005-u2os-72h-cp-a-au00044859_m07_s5_w1226d...  
4  taoe005-u2os-72h-cp-a-au00044859_i04_s5_w1f731...  

[5 rows x 23 columns]

<IPython.core.display.Javascript object>

After the filtering we are left with 1,918 unique plate-well combinations for each 9 fields of view are available leading to a total of 17,262 images. Surprisingly, those are 36 images (4 plate-well) more than what is described in the publication to be the final result of the preprocessing of the images. 

While our segmentation pipeline will differ from the ones the authors used to segment the nuclei, we can get also a first feeling of the dimension of the single-nuclei imaging dataset that we will be working with using the available metadata.

In [14]:
np.sum(list(filtered_metadata_orm["Image_Count_Nuclei"])), len(
    np.unique(filtered_metadata_orm["Image_Metadata_GeneSymbol"])
)

(1278881, 194)

<IPython.core.display.Javascript object>

The authors obtained roughly 1,28 million nuclei corresponding to ORF overexpression of 193 genes respectively the control condition. The total number of nuclei for each condition differs quite a bit, which is also due to the fact that we have more plate-well combinations for some genes than for others.

In [15]:
Counter(filtered_metadata_orm["Image_Metadata_GeneSymbol"])

Counter({'EIF4E': 90,
         'DKK1': 45,
         'ATF6': 135,
         'FURIN': 90,
         'MAPK8': 90,
         'CARD11': 135,
         'ATG16L1': 45,
         'TSC1': 90,
         'PAK1': 90,
         'XBP1': 180,
         'PRKAA1': 90,
         'MAP3K9': 45,
         'IKBKE': 90,
         'TGFBR1': 180,
         'RIPK1': 45,
         'EMPTY': 1557,
         'PSENEN': 45,
         'BRAF': 135,
         'DVL1': 90,
         'PER1': 90,
         'EGLN1': 135,
         'MOS': 90,
         'BMPR1B': 180,
         'Luciferase': 360,
         'CCND1': 90,
         'DVL3': 45,
         'TBK1': 90,
         'PIK3R1': 90,
         'PRKACA': 90,
         'NOTCH1': 135,
         'DEPTOR': 90,
         'TRAF6': 90,
         'PRKCZ': 180,
         'PRKACB': 135,
         'PRKACG': 135,
         'DIABLO': 45,
         'CHUK': 90,
         'LRPPRC': 45,
         'SLIRP': 90,
         'MKNK1': 45,
         'RAF1': 135,
         'GLI1': 45,
         'CYLD': 90,
         'JAK2': 90,
         'PKI

<IPython.core.display.Javascript object>

---

## 3. Save filtered image data

Previously we filtered the metadata to exclude images not passing the quality control standards established by Rohban et al. (2017). We will now use this subset and store the Hoechst(DNA) stained images individually. This is because a) we can drastically reduce the size of the overall dataset in terms of storage requirements and going forward we are only interested in this staining.

In [16]:
output_dir = "../data/images/filtered"
input_dir = "../data/images/data"
plate_col = "Image_Metadata_Plate"
illum_file_col = "Image_FileName_IllumHoechst"

# save_filtered_images(
#     filtered_metadata_orm,
#     input_dir=input_dir,
#     output_dir=output_dir,
#     plate_col=plate_col,
#     illum_file_col=illum_file_col,
# )

<IPython.core.display.Javascript object>

---

## 4. Nuclei segmentation

After the previously described filtering steps we will now turn to the segmentation of the invidiual nuclei as we will be working on single nuclei images. In addition to the raw images the authors also provide contour lines for the individual nuclei as segmented by their computational pipeline. We will use those and hence use their segmentation to save us from finetuning a segmentation pipeline as well as in order to be consistent with their preprocessing pipeline.

### 4.1. Visualize the segmentation

Before we use the given data to segment the individual nuclei we first provide some example of the segmentation to show that it accurately captures the nuclei. To this end, we will randomly sample n=10 images from each plate and overlay the segmentation based on the provided contour lines with the raw input image to show the segmentation.

In [17]:
n = 10
image_input_dir = "../data/images/filtered"
outline_input_dir = "../data/images/segmented_outlines"

(
    image_sample_datasets,
    segmentation_sample_datasets,
) = create_sample_segmentation_datasets(
    metadata=filtered_metadata_orm,
    image_input_dir=image_input_dir,
    outline_input_dir=outline_input_dir,
    n=n,
    plate_col="Image_Metadata_Plate",
    illum_file_col="Image_FileName_IllumHoechst",
    nuclei_outline_col="Image_FileName_NucleiOutlines",
    random_state=1234,
)

<IPython.core.display.Javascript object>

----

#### Segmented 41744

In [18]:
explore_segmentation_of_dataset(
    image_sample_datasets, segmentation_sample_datasets, dataset_id=0, cmap="seismic"
)

interactive(children=(IntSlider(value=0, description='plane', max=9), Output()), _dom_classes=('widget-interac…

<function src.utils.basic.visualization.explore_segmentation_of_dataset.<locals>.display_slice(plane=0)>

<IPython.core.display.Javascript object>

---

#### Segmented 41754

In [19]:
explore_segmentation_of_dataset(
    image_sample_datasets, segmentation_sample_datasets, dataset_id=1, cmap="seismic"
)

interactive(children=(IntSlider(value=0, description='plane', max=9), Output()), _dom_classes=('widget-interac…

<function src.utils.basic.visualization.explore_segmentation_of_dataset.<locals>.display_slice(plane=0)>

<IPython.core.display.Javascript object>

----

#### Segmented 41755

In [20]:
explore_segmentation_of_dataset(
    image_sample_datasets, segmentation_sample_datasets, dataset_id=2, cmap="seismic"
)

interactive(children=(IntSlider(value=0, description='plane', max=9), Output()), _dom_classes=('widget-interac…

<function src.utils.basic.visualization.explore_segmentation_of_dataset.<locals>.display_slice(plane=0)>

<IPython.core.display.Javascript object>

----

#### Segmented 41744

In [21]:
explore_segmentation_of_dataset(
    image_sample_datasets, segmentation_sample_datasets, dataset_id=3, cmap="seismic"
)

interactive(children=(IntSlider(value=0, description='plane', max=9), Output()), _dom_classes=('widget-interac…

<function src.utils.basic.visualization.explore_segmentation_of_dataset.<locals>.display_slice(plane=0)>

<IPython.core.display.Javascript object>

----

#### Segmented 41757

In [22]:
explore_segmentation_of_dataset(
    image_sample_datasets, segmentation_sample_datasets, dataset_id=4, cmap="seismic"
)

interactive(children=(IntSlider(value=0, description='plane', max=9), Output()), _dom_classes=('widget-interac…

<function src.utils.basic.visualization.explore_segmentation_of_dataset.<locals>.display_slice(plane=0)>

<IPython.core.display.Javascript object>

The above segmentation looks fine and hence we will use that pipeline to segment all images.

---

### 4.3. Segment and save images


During the segmentation we will ensure to maintain the available metadata including i.a. the gene targets, well and plate number for each nucleus by constructing a new metadata file. Those information will be important for our downstream analyses.

In [88]:
output_dir = "../data/images/single_nuclei_min_area_400"

<IPython.core.display.Javascript object>

In [ ]:
nuclei_metadata, image_metadata = segment_images_and_get_nuclei_metadata(
    metadata=filtered_metadata_orm,
    image_input_dir=image_input_dir,
    outline_input_dir=outline_input_dir,
    output_dir=os.path.join(output_dir, "images"),
    plate_col="Image_Metadata_Plate",
    illum_file_col="Image_FileName_IllumHoechst",
    nuclei_outline_col="Image_FileName_NucleiOutlines",
    nuclei_count_col="Image_Count_Nuclei",
    min_area=400,
)
nuclei_metadata.to_csv(os.path.join(output_dir, "nuclei_metadata.csv"))

  0%|          | 0/17262 [00:00<?, ?it/s]

<ipython-input-92-b736e45bad1c>:35: UserWarning: Only one label was provided to `remove_small_objects`. Did you mean to use a boolean array?
  label_image = morphology.remove_small_objects(label_image, min_size=min_area)


In [ ]:
nuclei_metadata